src/Main.hs

```diff
@@ -14,15 +14,143 @@
+
+white :: PixelRGB8
+white = PixelRGB8 255 255 255
+
+black :: PixelRGB8
+black = PixelRGB8 0 0 0
+
+redMaterial :: Material
+redMaterial = Material red
+
+spheres :: [Shape]
+spheres = [Sphere (Vector (-3) 3.5 (-8)) 3, Sphere (Vector 3 3.5 (-8)) 1]
+
+objects :: [Object]
+objects = [Object 
+                (Sphere (Vector (-3) 3.5 (-8)) 3) 
+                (Material red),
+           Object
+                (Sphere (Vector 3 3.5 (-8)) 1)
+                (Material green)]
+
+lights :: [Light]
+lights = [PointLight (Vector (-30) (-10) 20) ]
+
+camera :: Camera
+camera = Camera 45 (Vector 0 1.8 10) (Vector 0 3 0)
+
+scene :: Scene
+scene = Scene spheres lights camera
+
+main :: IO()
+main = writePng "output.png" $ generateImage Main.trace sceneWidth sceneHeight
+
+minWithEmpty :: [Double] -> Double
+minWithEmpty [] = -1
+minWithEmpty list = minimum list
+
+trace x y 
+    | isNothing intersection = white
+    | otherwise = getColorFromIntersection intersection
+    where intersection = closestIntersection (generateRay camera sceneWidth sceneHeight x y) objects
+
+getColorFromIntersection (Just ( _ , Object _ (Material color))) = color
+
+-- Generating rays, assuming distance to the image is 1 unit
+generateRay :: Camera -> Int -> Int -> Int -> Int -> Ray
+generateRay camera width height x y = 
+    let 
+         -- Vector from camera to lookAt point
+        eyeVector = buildEyeVector camera 
+        -- Vector in local right direction
+        rightVector = normalize (eyeVector `cross` Vector 0 1 0)
+        upVector = normalize (rightVector `cross` eyeVector)
+        -- Halves are taken to make right angles
+        halfFov = fovInRadians camera
+        -- This aspect ratio will be used, but are not the width and height of the camera
+        aspectRatio = h / w
+        halfWidth = tan halfFov
+        halfHeight = aspectRatio * halfWidth
+        cameraWidth = halfWidth * 2
+        cameraHeight = halfHeight * 2
+        pixelWidth = cameraWidth / (w - 1)
+        pixelHeight = cameraHeight / (h - 1)
+        scaledX = ((fromIntegral x * pixelWidth) - halfWidth)  `scalarMult` rightVector
+        scaledY = ((fromIntegral y * pixelHeight) - halfHeight) `scalarMult`  upVector 
+        orientation = normalize $ eyeVector `add` scaledX `add` scaledY
+    in  Ray (getPosition camera) orientation
+    where w = fromIntegral width
+          h = fromIntegral height
+
+buildEyeVector :: Camera -> Vector
+buildEyeVector (Camera _ location lookAt) = normalize (lookAt `sub` location)
+
+fovInRadians :: Camera -> Double
+fovInRadians (Camera fov _ _ ) = pi * (fov / 2) / 180
+
+getPosition :: Camera -> Vector
+getPosition (Camera _ position _ ) = position
+
+closestIntersection :: Ray -> [Object] -> Maybe (Double, Object)
+closestIntersection ray objects
+    | null intersections = Nothing
+    | otherwise = Just $ minimumBy tupleMinimum (map fromJust intersections)
+    where intersections = filter isJust $ map (intersects ray) objects
```

Comment:
so generally I use the where clauses to define functions that take args, whereas for fixed values I'll use let. This feels like it would make more sense as a let to me